File: src/domain/service/contest.ts
import { ContestsRepository } from "../../repository/contestRepository.js";
import { Contest } from "../contest.js";

export class ContestService {
  private readonly repository: ContestsRepository;
  constructor(repository: ContestsRepository) {
    this.repository = repository;
  }
  async Exists(d: Contest): Promise<boolean> {
    const all = await this.repository.findAll();
    if (all.isFailure()) {
      return true;
    }

    const iv = all.value.map((v) => {
      return [v.id, v.title];
    });

    return iv
      .map((v) => {
        return v[0] === d.id || v[1] === d.title;
      })
      .includes(true);
  }
}

File: pkg/domain/service/contest.go
package service

import (
	"github.com/poporonnet/kojsx-backend/pkg/domain"
	"github.com/poporonnet/kojsx-backend/pkg/repository"
)

type ContestService struct {
	contestRepository repository.ContestRepository
}

func NewContestService(repository repository.ContestRepository) *ContestService {
	return &ContestService{contestRepository: repository}
}

func (s *ContestService) IsExists(p domain.Contest) bool {
	// 重複判定: ID/Title
	i, _ := s.contestRepository.FindContestByID(p.GetID())
	t, _ := s.contestRepository.FindContestByTitle(p.GetTitle())
	if i == nil && t == nil {
		return false
	}

	return true
}
